src/control/cmd/dmg/pretty/firmware_test.go

```diff
@@ -47,6 +47,201 @@ func makeHostSet(t *testing.T, hosts string) *hostlist.HostSet {
+							ChannelPosition: 1,
+						},
+						Error: errors.New("test error"),
+					},
+					{
+						Module: storage.ScmModule{
+							UID:             "Device4",
+							PhysicalID:      4,
+							Capacity:        67890,
+							SocketID:        2,
+							ControllerID:    2,
+							ChannelID:       2,
+							ChannelPosition: 1,
+						},
+						Error: errors.New("test error"),
+					},
+				},
+			},
+			expPrintStr: `
+-----
+host1
+-----
+  Active Version: V100
+  Staged Version: V101
+  Maximum Firmware Image Size: 12 KiB
+  Last Update Status: Staged
+    2 devices
+-----
+host1
+-----
+  Error: test error
+    UID:Device3 PhysicalID:3 Capacity:66 KiB Location:(socket:1 memctrlr:2 chan:2 pos:1)
+    UID:Device4 PhysicalID:4 Capacity:66 KiB Location:(socket:2 memctrlr:2 chan:2 pos:1)
+`,
+		},
+		"multiple hosts": {
+			fwMap: control.HostSCMQueryMap{
+				"host1": []*control.SCMQueryResult{
+					{
+						Module: storage.ScmModule{
+							UID:             "Device1",
+							PhysicalID:      1,
+							Capacity:        (1 << 31),
+							SocketID:        1,
+							ControllerID:    2,
+							ChannelID:       3,
+							ChannelPosition: 5,
+						},
+						Error: errors.New("shared error"),
+					},
+				},
+				"host2": []*control.SCMQueryResult{
+					{
+						Module: storage.ScmModule{
+							UID:             "Device2",
+							PhysicalID:      2,
+							Capacity:        (1 << 30),
+							SocketID:        6,
+							ControllerID:    7,
+							ChannelID:       8,
+							ChannelPosition: 9,
+						},
+						Error: errors.New("shared error"),
+					},
+					{
+						Module: storage.ScmModule{
+							UID:             "Device3",
+							PhysicalID:      3,
+							Capacity:        (1 << 32),
+							SocketID:        1,
+							ControllerID:    2,
+							ChannelID:       2,
+							ChannelPosition: 1,
+						},
+						Info: &storage.ScmFirmwareInfo{
+							ActiveVersion:     "B300",
+							StagedVersion:     "",
+							ImageMaxSizeBytes: (1 << 20),
+							UpdateStatus:      storage.ScmUpdateStatusFailed,
+						},
+					},
+					{
+						Module: storage.ScmModule{
+							UID:             "Device4",
+							PhysicalID:      4,
+							Capacity:        (1 << 32),
+							SocketID:        1,
+							ControllerID:    2,
+							ChannelID:       3,
+							ChannelPosition: 1,
+						},
+						Info: &storage.ScmFirmwareInfo{
+							ActiveVersion:     "B300",
+							StagedVersion:     "",
+							ImageMaxSizeBytes: (1 << 20),
+							UpdateStatus:      storage.ScmUpdateStatusFailed,
+						},
+					},
+				},
+			},
+			expPrintStr: `
+-----
+host2
+-----
+  Active Version: B300
+  Staged Version: N/A
+  Maximum Firmware Image Size: 1.0 MiB
+  Last Update Status: Failed
+    2 devices
+---------
+host[1-2]
+---------
+  Error: shared error
+    UID:Device1 PhysicalID:1 Capacity:2.0 GiB Location:(socket:1 memctrlr:2 chan:3 pos:5)
+    UID:Device2 PhysicalID:2 Capacity:1.0 GiB Location:(socket:6 memctrlr:7 chan:8 pos:9)
```

Comment:
Hmm. This is still kind of difficult to digest, IMO. What about something like this?
```
Errors:
  Hosts     Error        Device:PhyID:Socket:Ctrl:Chan:Pos
  -----     -----        ---------------------------------
  host[1-2] shared error Device1:1:1:2:3:5
  host[1-2] shared error Device2:2:6:7:8:9

-----
host1
-----
  Firmware status for 2 devices:
    Active Version: V100
    Staged Version: V101
    Maximum Firmware Image Size: 12 KiB
    Last Update Status: Staged
```

I don't love repeating the hostrange/error on multiple rows, but I think the important piece of information is the device identifier. I also think this format will be more useful when we have heterogeneous failures across multiple hosts rather than these artificially-neat failures involving the same modules and same errors across multiple hosts.

A tweaked example of that plus moving the variable-length error string to the end might look something like this:
```
Errors:
  (Device: UID:PhysID:Socket:Controller:Channel:Position)
  Hosts Device            Error
  ----- ------            -----
  host1 Device1:1:1:2:3:5 error one
  host2 Device2:2:6:7:8:9 error two

-----
host1
-----
  Firmware status for 2 devices:
    Active Version: V100
    Staged Version: V101
    Maximum Firmware Image Size: 12 KiB
    Last Update Status: Staged
```

Reply:
Good suggestion. Reworking.